### PR TITLE
[Pallas] Introduce make_kernel_from_pallas 

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This pull request introduces make_kernel_from_pallas API which is the top level API to interact with the Pallas integration. It takes a pallas_call wrapper and than make it a custom pytorch op.

Test Plan:
python test/test_pallas.py

### CODE REVIEW

File: torch_xla/experimental/custom_kernel.py [hunk 1]
```diff
@@ -56,3 +62,50 @@ def _extract_backend_config(
       if op.name == "stablehlo.custom_call":
         return op.backend_config.value
   return None
+
+
+def convert_torch_dtype_to_jax(dtype: torch.dtype) -> jnp.dtype:
```
Comment (JackCaoG) [inline torch_xla/experimental/custom_kernel.py:67]: @qihqi do we already have such converstation somewhere in torchxla2?

File: test/test_pallas.py [hunk 2]
```diff
@@ -135,6 +135,48 @@ def add_vectors(x: jax.Array, y: jax.Array) -> jax.Array:
...
+      return pl.pallas_call(
+          add_vectors_kernel, out_shape=jax.ShapeDtypeStruct(x.shape,
+                                                             x.dtype))(x, y)
+
+    from torch_xla.experimental.custom_kernel import make_kernel_from_pallas
+    pt_kernel = make_kernel_from_pallas(add_vectors, lambda x, y:
+                                        (x.shape, x.dtype))
+
+    dtypes = [torch.float32, torch.float
+             ]  # TODO: torch.float64, torch.bfloat16, torch.float16 don't work.
```
Comment (JackCaoG) [inline test/test_pallas.py:162]: why bf16 won't work?

Comment (alanwaketan) [reply test/test_pallas.py:162]: Mosaic complaints. Need to dig more into it.

File: test/test_pallas.py [hunk 3]
```diff
@@ -135,6 +135,48 @@ def add_vectors(x: jax.Array, y: jax.Array) -> jax.Array:
     # the most important fields are present.
     self.assertIn("custom_call_config", payload)
 
+  @unittest.skipIf(xr.device_type() != 'TPU', "This test only works on TPU.")
+  # TODO: This test cannot be ran individually, let's fix it.
+  def test_tpu_custom_call_pallas_wrap_add_payload(self):
+    import jax
+    import jax.numpy as jnp
+    import jax._src.pallas.mosaic.pallas_call_registration
```
Comment (JackCaoG) [inline test/test_pallas.py:143]: seems like this is repeated on multiple tests, maybe just move to the top?

**Review by JackCaoG — COMMENTED** [review]
Do you need this pr in 2.3?

**Comment by alanwaketan** [comment]
> Do you need this pr in 2.3?

Yea, will also need a couple for the TODOs.

File: test/test_pallas.py (hunk 3 above)
Comment (alanwaketan) [reply test/test_pallas.py:143]: There is a compatibility issue where jax will try to lock tpu devices if we import them before any pt/xla computations... I will need to resolve that...

File: test/test_pallas.py [hunk 4]
```diff
@@ -135,6 +135,48 @@ def add_vectors(x: jax.Array, y: jax.Array) -> jax.Array:
     # the most important fields are present.
     self.assertIn("custom_call_config", payload)
 
+  @unittest.skipIf(xr.device_type() != 'TPU', "This test only works on TPU.")
+  # TODO: This test cannot be ran individually, let's fix it.
+  def test_tpu_custom_call_pallas_wrap_add_payload(self):
+    import jax
```
Comment (miladm) [inline test/test_pallas.py:141]: I am conerned JAX-based tests cause failures due to `libtpu` version inconsistencies, and in turn CI hiccups. How do we resolve this concern?

Comment (alanwaketan) [reply test/test_pallas.py:141]: That's resolved in the last PR: https://github.com/pytorch/xla/pull/6696

File: torch_xla/experimental/custom_kernel.py [hunk 5]
```diff
@@ -56,3 +62,50 @@ def _extract_backend_config(
...
+  elif dtype == torch.int8:
+    return jnp.int8
+  elif dtype == torch.uint8:
+    return jnp.uint8
+  else:
+    raise ValueError(f"Unsupported dtype: {dtype}")
+
+
+def make_kernel_from_pallas(kernel: Callable, output_shape_dtype_fn: Callable):
+  # TODO: Maybe we can cache the payload for the same input.
```
[the remark this answers is not in the document]
Comment (alanwaketan) [reply torch_xla/experimental/custom_kernel.py:91]: Right, the cache itself should deal with the dynamism.

**Review by JackCaoG — APPROVED** [review]
I still think we should refactor `convert_torch_dtype_to_jax` and invesgate bf16(which I assume most people will use), approve to unblock.

**Comment by alanwaketan** [comment]
> I still think we should refactor `convert_torch_dtype_to_jax` and invesgate bf16(which I assume most people will use), approve to unblock.

Yea, for sure. Let me follow up with that.

**Pull request merged**